core/src/main/scala/com/evolutiongaming/kafka/flow/registry/EntityRegistry.scala

```diff
@@ -0,0 +1,87 @@
+  * When the entity is removed from the in-memory cache, this association is also supposed to be removed.
+  *
+  * There are three pre-defined implementations currently provided:
+  *   - `EntityRegistry.empty` for no-op implementation
+  *   - `EntityRegistry.const` for immutable constant data
+  *   - `EntityRegistry.memory`, fully functional in-memory registry
+  * @tparam K entity key
+  * @tparam S entity state
+  */
+trait EntityRegistry[F[_], K, S] {
+
+  /** Registers an association between an entity's key and the function to obtain current state of it.
+    * This is used internally by the library and shouldn't be called by the library users.
+    * @param key entity key
+    * @param state computation (effectively function) to obtain the current value of the entity
+    * @return `Resource` which acquisition registers the entity and releasing removes it from the registry
+    */
+  def register(key: K, state: F[Option[S]]): Resource[F, Unit]
+
+  /** Returns current state of the entity by key (or None if there's no such entity or it has empty state)
+    * @param key entity key
+    * @return `Some` when the entity is present and has non-empty state, None otherwise
+    */
+  def get(key: K): F[Option[S]]
+
+  /** Returns keys and states of all currently registered in-memory entities.
+    * Filters out entities having None as state
+    */
+  def getAll: F[Map[K, S]]
+}
+
+object EntityRegistry {
+  def empty[F[_], K, S](implicit F: Applicative[F]): EntityRegistry[F, K, S] = new EmptyEntityRegistry[F, K, S]()
+
+  def memory[F[_], K, S](implicit F: Monad[F], R: Ref.Make[F]): F[EntityRegistry[F, K, S]] =
+    R.refOf(Map.empty[K, F[Option[S]]]).map(ref => new InMemoryEntityRegistry[F, K, S](ref))
+
+  def const[F[_], K, S](values: Map[K, S])(implicit F: Applicative[F]): EntityRegistry[F, K, S] =
+    new ConstEntityRegistry[F, K, S](values)
+
+  /** No-op registry, always returning None and an empty map on `get` and no-op on `register` */
+  final class EmptyEntityRegistry[F[_], K, S](implicit F: Applicative[F]) extends EntityRegistry[F, K, S] {
+    override def register(key: K, state: F[Option[S]]): Resource[F, Unit] = Resource.unit
+    override def get(key: K): F[Option[S]] = F.pure(none[S])
+    override def getAll: F[Map[K, S]] = F.pure(Map.empty)
+  }
+
+  /** Immutable registry with a no-op `register` method. `get` and `getAll` query the passed map */
+  final class ConstEntityRegistry[F[_], K, S](values: Map[K, S])(implicit F: Applicative[F])
+      extends EntityRegistry[F, K, S] {
+    override def register(key: K, state: F[Option[S]]): Resource[F, Unit] = Resource.unit
+    override def get(key: K): F[Option[S]] = F.pure(values.get(key))
+    override def getAll: F[Map[K, S]] = F.pure(values)
+  }
+
+  /** Mutable registry that keeps the state in a `Ref` */
+  final class InMemoryEntityRegistry[F[_], K, S](ref: Ref[F, Map[K, F[Option[S]]]])(implicit F: Monad[F])
```

Comment:
single `Ref` covering all entities - could be a bottleneck.

Reply:
Maybe we could first see how it does and if it's not enough then we can review our options (premature optimization and all that).  
What would you recommend as an alternative, would `com.evolutiongaming.scache.Cache#loading` be a good choice here in your opinion given how it's partitioned? I could introduce a separate implementation based on it next to this one so that users can choose themselves and not be bound to `Cache`'s implementations/possible bugs in any way.

Reply:
```
partitions: Partitions[Ref[Map[K, S]]
```


In this case I disagree about "premature optimization and all that" argument :)